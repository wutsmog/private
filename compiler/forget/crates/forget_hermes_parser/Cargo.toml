[package]
name = "forget_hermes_parser"
version = "0.1.0"
authors.workspace = true
description.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = { workspace = true }
forget_diagnostics = { workspace = true }
forget_estree = { workspace = true }
hermes = { workspace = true } 
juno_support = { workspace = true }
serde_json = { workspace = true }

[build-dependencies]
forget_estree_codegen = { workspace = true }