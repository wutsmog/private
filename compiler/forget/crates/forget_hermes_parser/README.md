# forget_hermes_parser

Wrapper around the Hermes parser that exposes parse results as a `forget_estree` AST.