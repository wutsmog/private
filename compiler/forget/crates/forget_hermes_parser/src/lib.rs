mod generated;
mod generated_extension;

use forget_diagnostics::Diagnostic;
use forget_estree::Program;
use generated_extension::{Context, FromHermes};
use hermes::parser::{HermesParser, ParserDialect, ParserFlags};
use hermes::utf::utf8_with_surrogates_to_string;
use juno_support::NullTerminatedBuf;

pub fn parse(source: &str, _file: &str) -> Result<Program, Vec<Diagnostic>> {
    let buf = NullTerminatedBuf::from_str_check(source);
    let result = HermesParser::parse(
        ParserFlags {
            dialect: ParserDialect::TypeScript,
            enable_jsx: true,
            store_doc_block: true,
            strict_mode: true,
        },
        &buf,
    );
    let mut cx = Context::new();
    if result.has_errors() {
        let error_messages = result.messages();
        return Err(error_messages
            .into_iter()
            .map(|diag| {
                let message = utf8_with_surrogates_to_string(diag.message.as_slice()).unwrap();
                Diagnostic::invalid_syntax(message, None)
            })
            .collect());
    }

    Ok(FromHermes::convert(&mut cx, result.root().unwrap()))
}
