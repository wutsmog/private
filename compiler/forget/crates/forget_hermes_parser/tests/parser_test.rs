use std::env;

use forget_hermes_parser::parse;
use insta::{assert_snapshot, glob};
use serde_json;

#[test]
fn fixtures() {
    glob!("fixtures/**.js", |path| {
        println!("fixture {}", path.to_str().unwrap());
        let input = std::fs::read_to_string(path).unwrap();
        let ast = parse(&input, path.to_str().unwrap()).unwrap();
        let output = serde_json::to_string_pretty(&ast).unwrap();
        let output = output.trim();
        assert_snapshot!(format!("Input:\n{input}\n\nOutput:\n{output}"));
    });
}
