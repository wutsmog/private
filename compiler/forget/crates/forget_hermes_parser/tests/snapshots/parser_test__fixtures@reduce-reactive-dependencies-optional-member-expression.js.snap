---
source: crates/forget_hermes_parser/tests/parser_test.rs
expression: "format!(\"Input:\\n{input}\\n\\nOutput:\\n{output}\")"
input_file: crates/forget_hermes_parser/tests/fixtures/reduce-reactive-dependencies-optional-member-expression.js
---
Input:
function Component(props) {
  const x = [];
  x.push(props.items?.length);
  x.push(props.items?.edges?.map?.(render)?.filter?.(Boolean) ?? []);
  return x;
}


Output:
{
  "type": "Program",
  "body": [
    {
      "type": "FunctionDeclaration",
      "id": {
        "type": "Identifier",
        "name": "Component",
        "typeAnnotation": null,
        "loc": null,
        "range": {
          "start": 9,
          "end": 18
        }
      },
      "params": [
        {
          "type": "Identifier",
          "name": "props",
          "typeAnnotation": null,
          "loc": null,
          "range": {
            "start": 19,
            "end": 24
          }
        }
      ],
      "body": {
        "type": "BlockStatement",
        "body": [
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "x",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 36,
                    "end": 37
                  }
                },
                "init": {
                  "type": "ArrayExpression",
                  "elements": [],
                  "loc": null,
                  "range": {
                    "start": 40,
                    "end": 42
                  }
                },
                "loc": null,
                "range": {
                  "start": 36,
                  "end": 42
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 30,
              "end": 43
            }
          },
          {
            "type": "ExpressionStatement",
            "expression": {
              "type": "CallExpression",
              "callee": {
                "type": "MemberExpression",
                "object": {
                  "type": "Identifier",
                  "name": "x",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 46,
                    "end": 47
                  }
                },
                "property": {
                  "type": "Identifier",
                  "name": "push",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 48,
                    "end": 52
                  }
                },
                "computed": false,
                "loc": null,
                "range": {
                  "start": 46,
                  "end": 52
                }
              },
              "arguments": [
                {
                  "type": "OptionalMemberExpression",
                  "object": {
                    "type": "MemberExpression",
                    "object": {
                      "type": "Identifier",
                      "name": "props",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 53,
                        "end": 58
                      }
                    },
                    "property": {
                      "type": "Identifier",
                      "name": "items",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 59,
                        "end": 64
                      }
                    },
                    "computed": false,
                    "loc": null,
                    "range": {
                      "start": 53,
                      "end": 64
                    }
                  },
                  "property": {
                    "type": "Identifier",
                    "name": "length",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 66,
                      "end": 72
                    }
                  },
                  "computed": false,
                  "optional": true,
                  "loc": null,
                  "range": {
                    "start": 53,
                    "end": 72
                  }
                }
              ],
              "loc": null,
              "range": {
                "start": 46,
                "end": 73
              }
            },
            "directive": null,
            "loc": null,
            "range": {
              "start": 46,
              "end": 74
            }
          },
          {
            "type": "ExpressionStatement",
            "expression": {
              "type": "CallExpression",
              "callee": {
                "type": "MemberExpression",
                "object": {
                  "type": "Identifier",
                  "name": "x",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 77,
                    "end": 78
                  }
                },
                "property": {
                  "type": "Identifier",
                  "name": "push",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 79,
                    "end": 83
                  }
                },
                "computed": false,
                "loc": null,
                "range": {
                  "start": 77,
                  "end": 83
                }
              },
              "arguments": [
                {
                  "type": "LogicalExpression",
                  "operator": "??",
                  "left": {
                    "type": "OptionalCallExpression",
                    "callee": {
                      "type": "OptionalMemberExpression",
                      "object": {
                        "type": "OptionalCallExpression",
                        "callee": {
                          "type": "OptionalMemberExpression",
                          "object": {
                            "type": "OptionalMemberExpression",
                            "object": {
                              "type": "MemberExpression",
                              "object": {
                                "type": "Identifier",
                                "name": "props",
                                "typeAnnotation": null,
                                "loc": null,
                                "range": {
                                  "start": 84,
                                  "end": 89
                                }
                              },
                              "property": {
                                "type": "Identifier",
                                "name": "items",
                                "typeAnnotation": null,
                                "loc": null,
                                "range": {
                                  "start": 90,
                                  "end": 95
                                }
                              },
                              "computed": false,
                              "loc": null,
                              "range": {
                                "start": 84,
                                "end": 95
                              }
                            },
                            "property": {
                              "type": "Identifier",
                              "name": "edges",
                              "typeAnnotation": null,
                              "loc": null,
                              "range": {
                                "start": 97,
                                "end": 102
                              }
                            },
                            "computed": false,
                            "optional": true,
                            "loc": null,
                            "range": {
                              "start": 84,
                              "end": 102
                            }
                          },
                          "property": {
                            "type": "Identifier",
                            "name": "map",
                            "typeAnnotation": null,
                            "loc": null,
                            "range": {
                              "start": 104,
                              "end": 107
                            }
                          },
                          "computed": false,
                          "optional": true,
                          "loc": null,
                          "range": {
                            "start": 84,
                            "end": 107
                          }
                        },
                        "arguments": [
                          {
                            "type": "Identifier",
                            "name": "render",
                            "typeAnnotation": null,
                            "loc": null,
                            "range": {
                              "start": 110,
                              "end": 116
                            }
                          }
                        ],
                        "optional": true,
                        "loc": null,
                        "range": {
                          "start": 84,
                          "end": 117
                        }
                      },
                      "property": {
                        "type": "Identifier",
                        "name": "filter",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 119,
                          "end": 125
                        }
                      },
                      "computed": false,
                      "optional": true,
                      "loc": null,
                      "range": {
                        "start": 84,
                        "end": 125
                      }
                    },
                    "arguments": [
                      {
                        "type": "Identifier",
                        "name": "Boolean",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 128,
                          "end": 135
                        }
                      }
                    ],
                    "optional": true,
                    "loc": null,
                    "range": {
                      "start": 84,
                      "end": 136
                    }
                  },
                  "right": {
                    "type": "ArrayExpression",
                    "elements": [],
                    "loc": null,
                    "range": {
                      "start": 140,
                      "end": 142
                    }
                  },
                  "loc": null,
                  "range": {
                    "start": 84,
                    "end": 142
                  }
                }
              ],
              "loc": null,
              "range": {
                "start": 77,
                "end": 143
              }
            },
            "directive": null,
            "loc": null,
            "range": {
              "start": 77,
              "end": 144
            }
          },
          {
            "type": "ReturnStatement",
            "argument": {
              "type": "Identifier",
              "name": "x",
              "typeAnnotation": null,
              "loc": null,
              "range": {
                "start": 154,
                "end": 155
              }
            },
            "loc": null,
            "range": {
              "start": 147,
              "end": 156
            }
          }
        ],
        "loc": null,
        "range": {
          "start": 26,
          "end": 158
        }
      },
      "generator": false,
      "async": false,
      "loc": null,
      "range": {
        "start": 0,
        "end": 158
      },
      "loc": null,
      "range": {
        "start": 0,
        "end": 158
      }
    }
  ],
  "sourceType": "script",
  "loc": null,
  "range": {
    "start": 0,
    "end": 158
  }
}
